content/typeclasses-syntax.md

```diff
@@ -0,0 +1,657 @@
+
+```scala
+  // Simple typeclass
+  given intOrd: Ord[Int] with
+    def compare(x: Int, y: Int) = ...
+
+  // Parameterized typeclass
+  given listOrd[A: Ord]: Ord[List[A]] with
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Typeclass with using clause
+  given listOrd[A](using Ord[A]): Ord[List[A]] with
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Simple alias
+  given intOrd: Ord[Int] = IntOrd()
+
+  // Parameterized alias
+  given listOrd[A: Ord]: Ord[List[A]] =
+    ListOrd[A]()
+
+  // Alias with using clause
+  given listOrd[A](using Ord[A]): Ord[List[A]] =
+    ListOrd[A]()
+
+  // Concrete class instance
+  given context: Context with {}
+
+  // Abstract or deferred given
+  given context: Context
+
+  // By-name given
+  given context[DummySoItsADef]: Context = curCtx
+```
+**Notes:**
+
+ - The interior `:` looks less out of place with names. One can clearly see the lineage where it came from.
+
+**Proposal, Named**
+
+```scala
+  // Simple typeclass
+  given Ord[Int] as intOrd:
+    def compare(x: Int, y: Int) = ...
+
+  // Parameterized typeclass
+  given [A: Ord] => Ord[List[A]] as listOrd:
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Typeclass with using clause
+  given [A](using Ord[A]) => Ord[List[A]] as listOrd:
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Simple alias
+  given Ord[Int] as intOrd = IntOrd()
+
+  // Parameterized alias
+  given [A: Ord] => Ord[List[A]] as listOrd =
+    ListOrd[A]()
+
+  // Alias with using clause
+  given [A](using Ord[A]) => Ord[List[A]] as listOrd =
+    ListOrd[A]()
+
+  // Concrete class instance
+  given Context as context
+
+  // Abstract or deferred given
+  given Context as context = deferred
+
+  // By-name given
+  given => Context as context = curCtx
+```
+**Notes:**
+
+ - The `as` takes some getting used to. However, it is in line with the concept that names
+   should come last since the important thing here is the type hat's implemented.
+
+**Proposal, Named in Current Style**
+
+As an alternative, here is a version of new style given, but using the current `id:` for syntax for optional names.
+
+```scala
+  // Simple typeclass
+  given intOrd: Ord[Int]:
+    def compare(x: Int, y: Int) = ...
+
+  // Parameterized typeclass
+  given listOrd: [A: Ord] => Ord[List[A]]:
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Typeclass with using clause
+  given listOrd: [A](using Ord[A]) => Ord[List[A]]:
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Simple alias
+  given intOrd: Ord[Int] = IntOrd()
+
+  // Parameterized alias
+  given listOrd: [A: Ord] => Ord[List[A]] =
+    ListOrd[A]()
+
+  // Alias with using clause
+  given listOrd: [A](using Ord[A]) => Ord[List[A]] =
+    ListOrd[A]()
+
+  // Concrete class instance
+  given context: Context // this would be a change of meaning from abstract given
+
+  // Abstract or deferred given
+  given context: Context = deferred
+
+  // By-name given
+  given context: => Context = curCtx
```

Comment:
I'm not sure if this is the right place to post this, or whether it's perhaps too late. But I was wondering, whether we could have an alternative proposal which looks like this.

The benefit I see it this is that it keeps the syntax very regular and intentionally resembles definition of methods (but with `given` instead of `def`).
```suggestion
  // Simple typeclass
  given intOrd: Ord[Int]:
    def compare(x: Int, y: Int) = ...

  // Parameterized typeclass
  given listOrd[A: Ord]: Ord[List[A]]:
    def compare(x: List[A], y: List[A]) = ...

  // Typeclass with using clause
  given listOrd[A](using Ord[A]): Ord[List[A]]:
    def compare(x: List[A], y: List[A]) = ...

  // Simple alias
  given intOrd: Ord[Int] = IntOrd()

  // Parameterized alias
  given listOrd[A: Ord]: Ord[List[A]] =
    ListOrd[A]()

  // Alias with using clause
  given listOrd[A](using Ord[A]): Ord[List[A]] =
    ListOrd[A]()

  // Concrete class instance
  given context: Context // this would be a change of meaning from abstract given

  // Abstract or deferred given
  given context: Context = deferred

  // By-name given
  given context: => Context = curCtx
```